controllers/cloudstackmachine_controller.go

```diff
@@ -137,19 +137,20 @@ func (r *CloudStackMachineReconciliationRunner) ConsiderAffinity() (ctrl.Result,
 // SetFailureDomainOnCSMachine sets the failure domain the machine should launch in.
 func (r *CloudStackMachineReconciliationRunner) SetFailureDomainOnCSMachine() (retRes ctrl.Result, reterr error) {
 	if r.ReconciliationSubject.Spec.FailureDomainName == "" {
-
+		var name string
 		if r.CAPIMachine.Spec.FailureDomain != nil &&
 			(util.IsControlPlaneMachine(r.CAPIMachine) || // Is control plane machine -- CAPI will specify.
 				*r.CAPIMachine.Spec.FailureDomain != "") { // Or potentially another machine controller specified.
+			name = *r.CAPIMachine.Spec.FailureDomain
 			r.ReconciliationSubject.Spec.FailureDomainName = *r.CAPIMachine.Spec.FailureDomain
 		} else { // Not a control plane machine. Place randomly.
 			randNum := (rand.Int() % len(r.CSCluster.Spec.FailureDomains)) // #nosec G404 -- weak crypt rand doesn't matter here.
-			name := r.CSCluster.Spec.FailureDomains[randNum].Name
-			if !strings.HasSuffix(name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
-				name = name + "-" + r.CAPICluster.Name
-			}
-			r.ReconciliationSubject.Spec.FailureDomainName = name
+			name = r.CSCluster.Spec.FailureDomains[randNum].Name
+		}
+		if !strings.HasSuffix(name, "-"+r.CAPICluster.Name) { // Add cluster name suffix if missing.
+			name = name + "-" + r.CAPICluster.Name
```

Comment:
If the failure domain name from capi machine does not end with the cluster name then the capi machine's failure domain and the CS machine's failure domain will be different. Is this okay?

Reply:
all newer version failure domain has cluster name as appended. 
Those mis-matched failure domains will cleared up. 

In the first upgrade, failure domain will be recreated with new names.

Reply:
Does this mean we need to re-evaluate this condition? https://github.com/maxdrib/eks-anywhere/blob/capc-v1beta2/pkg/api/v1alpha1/cloudstackdatacenterconfig_webhook.go#L101. I was assuming that we can identify the "v1beta1 -> v1beta2 CAPC upgrade" scenario and allow the webhook to skip some validation for it

Reply:
last 8 chars zone ID - cluster name